src/Features/Core/Portable/Snippets/SnippetProviders/AbstractIfSnippetProvider.cs

```diff
@@ -0,0 +1,130 @@
+        {
+            var semanticModel = await document.ReuseExistingSpeculativeModelAsync(position, cancellationToken).ConfigureAwait(false);
+
+            var syntaxContext = document.GetRequiredLanguageService<ISyntaxContextService>().CreateContext(document, semanticModel, position, cancellationToken);
+            return syntaxContext.IsStatementContext || syntaxContext.IsGlobalStatementContext;
+        }
+
+        protected override Task<ImmutableArray<TextChange>> GenerateSnippetTextChangesAsync(Document document, int position, CancellationToken cancellationToken)
+        {
+            var snippetTextChange = GenerateSnippetTextChange(document, position);
+            return Task.FromResult(ImmutableArray.Create(snippetTextChange));
+        }
+
+        private static TextChange GenerateSnippetTextChange(Document document, int position)
+        {
+            var generator = SyntaxGenerator.GetGenerator(document);
+            var ifStatement = generator.IfStatement(generator.TrueLiteralExpression(), Array.Empty<SyntaxNode>());
+
+            return new TextChange(TextSpan.FromBounds(position, position), ifStatement.ToFullString());
+        }
+
+        protected override int? GetTargetCaretPosition(ISyntaxFactsService syntaxFacts, SyntaxNode caretTarget)
+        {
+            GetPartsOfIfStatement(caretTarget, out _, out _, out var statement);
+            return statement.SpanStart + 1;
+        }
+
+        protected override async Task<SyntaxNode> AnnotateNodesToReformatAsync(Document document,
+            SyntaxAnnotation findSnippetAnnotation, SyntaxAnnotation cursorAnnotation, int position, CancellationToken cancellationToken)
+        {
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var snippetExpressionNode = FindAddedSnippetSyntaxNode(root, position, syntaxFacts);
+            if (snippetExpressionNode is null)
+            {
+                return root;
+            }
+
+            var reformatSnippetNode = snippetExpressionNode.WithAdditionalAnnotations(findSnippetAnnotation, cursorAnnotation, Simplifier.Annotation, Formatter.Annotation);
+            return root.ReplaceNode(snippetExpressionNode, reformatSnippetNode);
+        }
+
+        protected override List<(string, List<TextSpan>)> GetRenameLocationsMap(SyntaxNode node, ISyntaxFacts syntaxFacts, CancellationToken cancellationToken)
+        {
+            var renameLocationsMap = new List<(string, List<TextSpan>)>();
+            GetPartsOfIfStatement(node, out _, out var condition, out var statement);
+
+            var list1 = new List<TextSpan>
+            {
+                // Need to add 1 to the span start because we are retrieving a block and want the cursor to appear inside the block.
+                new TextSpan(statement.SpanStart + 1, statement.Span.Length)
```

Comment:
i'm not sure how this is a rename location.  if you have `if (true) { }` then you're setting the rename space to `{[| } |]` (esp because you're still using the length of hte block without adjusting to moving the starting poitn forward.  I'm not sure how this is a rename location.

Reply:
It's a tab stop location, it is where we want the cursor to end up?